internal/hcsoci/devices.go

```diff
@@ -144,6 +158,75 @@ func handleAssignedDevicesWindows(ctx context.Context, vm *uvm.UtilityVM, annota
 	return resultDevs, closers, nil
 }
 
+// handleAssignedDevicesLCOW does all of the work to setup the hosting UVM, assign in devices
+// specified on the spec
+//
+// For LCOW, drivers must be installed before the target devices are assigned into the UVM so they
+// can be linked on arrival.
+func handleAssignedDevicesLCOW(ctx context.Context, vm *uvm.UtilityVM, annotations map[string]string, specDevs []specs.WindowsDevice) (resultDevs []specs.WindowsDevice, closers []resources.ResourceCloser, err error) {
+	defer func() {
+		if err != nil {
+			// best effort clean up allocated resources on failure
+			for _, r := range closers {
+				if releaseErr := r.Release(ctx); releaseErr != nil {
+					log.G(ctx).WithError(releaseErr).Error("failed to release container resource")
+				}
+			}
+			closers = nil
+			resultDevs = nil
+		}
+	}()
+
+	gpuPresent := false
+
+	// assign device into UVM and create corresponding spec windows devices
+	for _, d := range specDevs {
+		switch d.IDType {
+		case uvm.VPCIDeviceIDType, uvm.VPCIDeviceIDTypeLegacy, uvm.GPUDeviceIDType:
+			gpuPresent = d.IDType == uvm.GPUDeviceIDType
```

Comment:
`specDevs` aren't sorted in any way, right? If that's the case, then I think you might be overriding the value of `gpuPresent` if, e.g. your devices are in an order like: `uvm.GPUDeviceIDType` and then `uvm.VPCIDeviceIDTypeLegacy`, maybe doing something like this will help:
```
gpuPresent = gpuPresent || (d.IDType == uvm.GPUDeviceIDType)
```
This should make sure that you don't flip `gpuPresent` from `true` to `false`

Reply:
ah, good catch thank you